pkg/admission/admissiontimeout/timeoutadmission.go

```diff
@@ -0,0 +1,67 @@
+package admissiontimeout
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apiserver/pkg/admission"
+)
+
+type pluginHandlerWithTimeout struct {
+	name            string
+	admissionPlugin admission.Interface
+	timeout         time.Duration
+}
+
+var _ admission.ValidationInterface = &pluginHandlerWithTimeout{}
+var _ admission.MutationInterface = &pluginHandlerWithTimeout{}
+
+func (p pluginHandlerWithTimeout) Handles(operation admission.Operation) bool {
+	return p.admissionPlugin.Handles(operation)
+}
+
+func (p pluginHandlerWithTimeout) Admit(a admission.Attributes, o admission.ObjectInterfaces) error {
+	mutatingHandler, ok := p.admissionPlugin.(admission.MutationInterface)
+	if !ok {
+		return nil
+	}
+
+	admissionDone := make(chan struct{})
+	var admissionErr error
+	go func() {
+		utilruntime.HandleCrash()
+		defer close(admissionDone)
+		admissionErr = mutatingHandler.Admit(a, o)
+	}()
+
+	select {
+	case <-admissionDone:
+		return admissionErr
+	case <-time.After(p.timeout):
+		return errors.NewTimeoutError(fmt.Sprintf("admission plugin %q failed to complete mutation in %v", p.name, p.timeout), int(p.timeout.Seconds())+1)
+	}
+}
+
+func (p pluginHandlerWithTimeout) Validate(a admission.Attributes, o admission.ObjectInterfaces) error {
+	validatingHandler, ok := p.admissionPlugin.(admission.ValidationInterface)
+	if !ok {
+		return nil
+	}
+
+	admissionDone := make(chan struct{})
+	var admissionErr error
+	go func() {
+		utilruntime.HandleCrash()
```

Comment:
if we panic we will close `admissionDone` and return `nil` to the caller, am I right?

Reply:
> if we panic we will close `admissionDone` and return `nil` to the caller, am I right?

I hope so?  If we miss a few I'm not super worried about it.  Panic-ing admission is very bad.

Reply:
returning an `err` instead is not worth it? I think that would require to have a custom `recover` routine.
On the other hand, the good thing is that `HandleCrash` logs the error anyway so we will not lose it.